### PR TITLE
feat(Ruleticket): add Project assign to RuleTicket action

### PR DESCRIPTION
**Pull request by stonebuzz**
Add capability to assign ```Project``` from ```RuleTicket``` action. 

This also allows you to use this rule action as a rule criteria from ```SLA Level Escalation``` 

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !26077

### CODE REVIEW

File: tests/functionnal/SLM.php [hunk 1]
```diff
@@ -35,6 +35,7 @@
 
 namespace tests\units;
 
+use CommonITILObject;
```
Comment (trasher) [inline tests/functionnal/SLM.php:38]: Not required?

**Review by trasher — APPROVED** [review]

File: tests/functionnal/SLM.php (hunk 1 above)
Comment (stonebuzz) [reply tests/functionnal/SLM.php]: yes, it's a residue, my bad

File: src/RuleTicket.php [hunk 2]
```diff
@@ -288,17 +296,22 @@ public function executeActions($output, $params, array $input = [])
...
                         if ($action->fields["field"] === "assign_appliance") {
                             if (!array_key_exists("items_id", $output) || $output['items_id'] == '0') {
                                 $output["items_id"] = [];
                             }
                             $output["items_id"][Appliance::getType()][] = $value;
+                        } else if ($action->fields["field"] === "assign_project") {
+                            if (!array_key_exists("items_id", $output) || $output['items_id'] == '0') {
+                                $output["items_id"] = [];
+                            }
+                            $output["items_id"][Project::getType()][] = $value;
```
Comment (cedric-anne) [inline src/RuleTicket.php:309]:
```suggestion
                            if (!array_key_exists("_projects_id", $output)) {
                                $output["_projects_id"] = [];
                            }
                            $output["_projects_id"][] = $value;
```

File: src/RuleTicket.php [hunk 3]
```diff
@@ -461,6 +474,30 @@ public function executeActions($output, $params, array $input = [])
...
+                                  $regexvalue = $action->fields["value"];
+                            }
+
+                            if (!is_null($regexvalue)) {
+                                $projects = new Project();
+                                $target_projects = $projects->find(["name" => $regexvalue]);
+
+                                if ((!array_key_exists("items_id", $output) || $output['items_id'] == '0') && count($target_projects) > 0) {
+                                    $output["items_id"] = [];
+                                }
```
Comment (cedric-anne) [inline src/RuleTicket.php:493]:
```suggestion
                                if (!array_key_exists("_projects_id", $output) && count($target_projects) > 0) {
                                    $output["_projects_id"] = [];
                                }
```

File: src/RuleTicket.php [hunk 4]
```diff
@@ -461,6 +474,30 @@ public function executeActions($output, $params, array $input = [])
...
+
+                                foreach ($target_projects as $value) {
+                                    $output["projects_id"] = $value['id'];
```
Comment (cedric-anne) [inline src/RuleTicket.php:496]:
```suggestion
                                    $output["projects_id"][] = $value['id'];
```

File: src/RuleTicket.php [hunk 5]
```diff
@@ -858,6 +895,13 @@ public function getActions()
         $actions['assign_appliance']['force_actions']         = ['assign','regex_result', 'append'];
         $actions['assign_appliance']['appendto']              = 'items_id';
 
+        $actions['assign_project']['name']                  = _n('Associated element', 'Associated elements', Session::getPluralNumber()) . " : " . Project::getTypeName(1);
+        $actions['assign_project']['type']                  = 'dropdown';
+        $actions['assign_project']['table']                 = 'glpi_projects';
+        $actions['assign_project']['permitseveral']         = ['append'];
+        $actions['assign_project']['force_actions']         = ['assign','regex_result', 'append'];
+        $actions['assign_project']['appendto']              = 'items_id';
```
Comment (cedric-anne) [inline src/RuleTicket.php:903]: Not sure it is used.
```suggestion
        $actions['assign_project']['appendto']              = '_projects_id';
```

File: src/Ticket.php [hunk 6]
```diff
@@ -1760,6 +1760,17 @@ public function post_updateItem($history = 1)
                 ]
             );
         }
+
+        // Add linked project
+        $projects_id = $this->input['_projects_id'] ?? 0;
+        if ($projects_id) {
```
Comment (cedric-anne) [inline src/Ticket.php:1766]:
```suggestion
        // Add linked projects
        $projects_ids = $this->input['_projects_id'] ?? [];
        foreach ($projects_ids as $projects_id) {
```

File: src/Ticket.php [hunk 7]
```diff
@@ -2163,6 +2174,17 @@ public function post_addItem()
             ]);
         }
 
+        // Add linked project
+        $projects_id = $this->input['_projects_id'] ?? 0;
+        if ($projects_id) {
```
Comment (cedric-anne) [inline src/Ticket.php:2179]:
```suggestion
        // Add linked projects
        $projects_ids = $this->input['_projects_id'] ?? [];
        foreach ($projects_ids as $projects_id) {
```

File: src/RuleTicket.php [hunk 8]
```diff
@@ -858,6 +895,13 @@ public function getActions()
         $actions['assign_appliance']['force_actions']         = ['assign','regex_result', 'append'];
         $actions['assign_appliance']['appendto']              = 'items_id';
 
+        $actions['assign_project']['name']                  = _n('Associated element', 'Associated elements', Session::getPluralNumber()) . " : " . Project::getTypeName(1);
```
Comment (cedric-anne) [inline src/RuleTicket.php:898]:
```suggestion
        $actions['assign_project']['name']                  = Project::getTypeName(1);
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: src/RuleTicket.php [hunk 9]
```diff
@@ -246,14 +246,22 @@ public function executeActions($output, $params, array $input = [])
...
                             }
                             $output["items_id"][Appliance::getType()][] = $action->fields["value"];
                         }
 
+                        // special case of project
+                        if ($action->fields["field"] == "assign_project") {
+                            if (!array_key_exists("items_id", $output) || $output['items_id'] == '0') {
+                                $output["items_id"] = [];
+                            }
+                            $output["_projects_id"] = $action->fields["value"];
```
[the remark this answers is not in the document]
Comment (stonebuzz) [reply src/RuleTicket.php:262]: fixed

File: src/RuleTicket.php (hunk 2 above)
Comment (stonebuzz) [reply src/RuleTicket.php:309]: fixed

File: src/RuleTicket.php (hunk 4 above)
Comment (stonebuzz) [reply src/RuleTicket.php:496]: fixed

File: src/RuleTicket.php (hunk 5 above)
Comment (stonebuzz) [reply src/RuleTicket.php:903]: fixed

File: src/Ticket.php (hunk 6 above)
Comment (stonebuzz) [reply src/Ticket.php:1766]: fixed

File: src/Ticket.php (hunk 7 above)
Comment (stonebuzz) [reply src/Ticket.php:2179]: fixed

File: src/RuleTicket.php (hunk 8 above)
Comment (stonebuzz) [reply src/RuleTicket.php:898]: fixed

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**